pkg/guestagent/events/containerd_linux.go

```diff
@@ -0,0 +1,365 @@
+	return &ContainerdEventMonitor{
+		socketPaths: socketPaths,
+	}
+}
+
+func (c *ContainerdEventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event, errCh chan error) {
+	const defaultRetryDelay = 2
+	retryDelay := defaultRetryDelay
+	var wg sync.WaitGroup
+	for _, socket := range c.socketPaths {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(time.Duration(retryDelay) * time.Second):
+					logrus.Debugf("attempting to connect to containerd socket %s after: %d", socket, retryDelay)
+					retryDelay = defaultRetryDelay
+				}
+				info, err := os.Stat(socket)
+				if err != nil {
+					if os.IsNotExist(err) {
+						logrus.Warnf("containerd socket %s does not exist", socket)
+					} else {
+						logrus.Errorf("failed to stat containerd socket %s: %v", socket, err)
+					}
+					retryDelay = 5
+					continue
+				}
+				if info.IsDir() {
+					errCh <- fmt.Errorf("containerd socket path %s is a directory", socket)
+					retryDelay = 15
+					continue
+				}
+				client, err := containerd.New(socket, containerd.WithDefaultNamespace(containerdNamespace.Default))
+				if err != nil {
+					logrus.Warnf("failed to create client for socket %s: %v", socket, err)
+					continue
+				}
+				logrus.Debugf("created containerd client for socket %s", socket)
+				clientCtx, cancel := context.WithTimeout(ctx, defaultSocketTimeout)
+				serving, serveErr := client.IsServing(clientCtx)
+				cancel()
+				if serveErr != nil || !serving {
+					logrus.Warnf("containerd daemon not serving on socket %s: %v. Retrying in 5s...", socket, serveErr)
+					client.Close()
+					retryDelay = 5
+					continue
+				}
+				logrus.Infof("successfully connected to containerd on socket %s", socket)
+				if err := runMonitorClient(ctx, client, ch); err != nil {
+					logrus.Errorf("containerd port monitoring for socket: %s failed: %s", socket, err)
+					errCh <- err
+				}
+				client.Close()
+			}
+		}(socket)
+	}
+	wg.Wait()
+}
+
+func runMonitorClient(ctx context.Context, client *containerd.Client, ch chan *api.Event) error {
+	runningContainers := make(ipPortMap)
+	subscribeFilters := []string{
+		`topic=="/tasks/start"`,
+		`topic=="/containers/update"`,
+		`topic=="/tasks/exit"`,
+	}
+	msgCh, errCh := client.Subscribe(ctx, subscribeFilters...)
+
+	if err := initializeRunningContainers(ctx, client, ch, runningContainers); err != nil {
+		return fmt.Errorf("failed to initialize existing containers published ports: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancellation: %w", ctx.Err())
+
+		case err := <-errCh:
+			return fmt.Errorf("receiving container event failed: %w", err)
+
+		case envelope := <-msgCh:
+			logrus.Debugf("received an event: %+v", envelope.Topic)
+			switch envelope.Topic {
+			case "/tasks/start":
+				startTask := &events.TaskStart{}
```

Comment:
Why did you change the word order?

```suggestion
				taskStart := &events.TaskStart{}
```

Also use consistently in the error messages below.